### PR TITLE
Feature/901adopt standard error handling quoting service

### PR DESCRIPTION
**Pull request by ndonnan**
Add standard error handling to the quoting service

### CODE REVIEW

File: src/model/quotes.js [hunk 1]
```diff
@@ -66,13 +66,10 @@ class QuotesModel {
 
     // make sure initiator is the PAYER party
     if (quoteRequest.transactionType.initiator !== 'PAYER') {
-      throw new Errors.FSPIOPError(quoteRequest, 'Only PAYER initiated transactions are supported',
-        fspiopSource, Errors.ApiErrorCodes.SERVER_ERROR)
+      throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(2000, 'Only PAYER initiated transactions are supported', null, fspiopSource)
```
Comment (rmothilal) [inline src/model/quotes.js]: why 2000? is there an enum for the error code?

File: src/model/quotes.js [hunk 2]
```diff
@@ -115,9 +112,7 @@ class QuotesModel {
         // fail fast on duplicate
         if (dupe.isDuplicateId && (!dupe.isResend)) {
           // same quoteId but a different request, this is an error!
-          throw new Errors.FSPIOPError(quoteRequest,
-            `Quote ${quoteRequest.quoteId} is a duplicate but hashes dont match`, fspiopSource,
-            Errors.ApiErrorCodes.MODIFIED_REQUEST)
+          throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(3106, `Quote ${quoteRequest.quoteId} is a duplicate but hashes dont match`, null, fspiopSource)
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 3]
```diff
@@ -279,20 +274,31 @@ class QuotesModel {
       let res
       try {
         res = await fetch(fullUrl, opts)
-      } catch (e) {
-        throw new Errors.FSPIOPError('Network error', `Network error forwarding quote request: ${e.stack || util.inspect(e)}`,
-          fspiopSource, Errors.ApiErrorCodes.DESTINATION_COMMUNICATION_ERROR)
+      } catch (err) {
+        throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(1001, `Network error forwarding quote request to ${fspiopDest}`, err, fspiopSource, [
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 4]
```diff
@@ -279,20 +274,31 @@ class QuotesModel {
...
+          { key: 'request', value: JSON.stringify(opts) }
+        ])
       }
       this.writeLog(`forwarding quote request ${quoteId} from ${fspiopSource} to ${fspiopDest} got response ${res.status} ${res.statusText}`)
 
       // handle non network related errors below
       if (!res.ok) {
-        throw new Errors.FSPIOPError(res.statusText, 'Got non-success response forwarding quote request',
-          fspiopSource, Errors.ApiErrorCodes.DESTINATION_COMMUNICATION_ERROR)
+        throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(1001, 'Got non-success response forwarding quote request', null, fspiopSource, [
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 5]
```diff
@@ -473,9 +478,8 @@ class QuotesModel {
       if (!endpoint) {
         // we didnt get an endpoint for the payee dfsp!
         // make an error callback to the initiator
-        return this.sendErrorCallback(new Errors.FSPIOPError(null,
-          `No FSPIOP_CALLBACK_URL_QUOTES found for quote ${quoteId} PAYER party`, fspiopSource, Errors.ApiErrorCodes.COMMUNICATION_ERROR),
-        quoteId)
+        const fspiopError = ErrorHandler.CreateFSPIOPErrorFromErrorCode(3201, `No FSPIOP_CALLBACK_URL_QUOTES found for quote ${quoteId} PAYER party`, null, fspiopSource)
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 6]
```diff
@@ -493,19 +497,30 @@ class QuotesModel {
       let res
       try {
         res = await fetch(fullUrl, opts)
-      } catch (e) {
-        throw new Errors.FSPIOPError('Network error', `Network error forwarding quote response: ${e.stack || util.inspect(e)}`,
-          fspiopSource, Errors.ApiErrorCodes.DESTINATION_COMMUNICATION_ERROR)
+      } catch (err) {
+        throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(1001, 'Network error forwarding quote response', err, fspiopSource, [
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 7]
```diff
@@ -493,19 +497,30 @@ class QuotesModel {
...
+          { key: 'method', value: opts.method },
+          { key: 'request', value: JSON.stringify(opts) }
+        ])
       }
       this.writeLog(`forwarding quote response got response ${res.status} ${res.statusText}`)
 
       if (!res.ok) {
-        throw new Errors.FSPIOPError(res.statusText, 'Got non-success response forwarding quote response',
-          fspiopSource, Errors.ApiErrorCodes.DESTINATION_COMMUNICATION_ERROR)
+        throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(1001, 'Got non-success response forwarding quote response', null, fspiopSource, [
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 8]
```diff
@@ -640,9 +649,7 @@ class QuotesModel {
       if (!endpoint) {
         // we didnt get an endpoint for the payee dfsp!
         // make an error callback to the initiator
-        throw new Errors.FSPIOPError(null,
-          `No FSPIOP_CALLBACK_URL_QUOTES found for quote GET ${quoteId}`, fspiopSource,
-          Errors.ApiErrorCodes.DESTINATION_FSP_ERROR)
+        throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(3201, `No FSPIOP_CALLBACK_URL_QUOTES found for quote GET ${quoteId}`, null, fspiopSource)
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 9]
```diff
@@ -659,40 +666,47 @@ class QuotesModel {
       let res
       try {
         res = await fetch(fullUrl, opts)
-      } catch (e) {
-        throw new Errors.FSPIOPError('Network error', `Network error forwarding quote get request: ${e.stack || util.inspect(e)}`,
-          fspiopSource, Errors.ApiErrorCodes.DESTINATION_COMMUNICATION_ERROR)
+      } catch (err) {
+        throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(1001, 'Network error forwarding quote get request', err, fspiopSource, [
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 10]
```diff
@@ -659,40 +666,47 @@ class QuotesModel {
...
+          { key: 'request', value: JSON.stringify(opts) }
+        ])
       }
       this.writeLog(`forwarding quote get request ${quoteId} from ${fspiopSource} to ${fspiopDest} got response ${res.status} ${res.statusText}`)
 
       // handle non network related errors below
       if (!res.ok) {
-        throw new Errors.FSPIOPError(res.statusText, 'Got non-success response forwarding quote get request',
-          fspiopSource, Errors.ApiErrorCodes.DESTINATION_COMMUNICATION_ERROR)
+        throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(1001, 'Got non-success response forwarding quote get request', null, fspiopSource, [
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 11]
```diff
@@ -707,54 +721,63 @@ class QuotesModel {
...
           'fspiop-http-method': 'PUT'
         }, true)
       }
       let res
       try {
         res = await axios.request(opts)
-      } catch (_err) {
-        throw new Error(`network error in sendErrorCallback: ${_err.message}`)
+      } catch (err) {
+        throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(1001, `network error in sendErrorCallback: ${err.message}`, err, fspiopSource, [
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: src/model/quotes.js [hunk 12]
```diff
@@ -707,54 +721,63 @@ class QuotesModel {
...
+          { key: 'destinationFsp', value: headers['fspiop-destination'] },
+          { key: 'method', value: opts.method },
+          { key: 'request', value: JSON.stringify(opts) }
+        ])
       }
       this.writeLog(`Error callback got response ${res.status} ${res.statusText}`)
 
       if (res.status !== 200) {
-        throw new Error('Got non-success response sending error callback')
+        throw ErrorHandler.CreateFSPIOPErrorFromErrorCode(1001, 'Got non-success response sending error callback', null, fspiopSource, [
```
Comment (rmothilal) [inline src/model/quotes.js]: Please use the enum for errorCode from the Error library

File: test/unit/model/quotes.test.js [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+      await quotesModel.validateQuoteRequest('dfsp1', {
+        transactionType: {
+          initiator: 'PAYEE'
+        }
+      })
+      test.fail('Expected an error to be thrown')
+      test.end()
+    } catch (err) {
+      test.ok(err instanceof FSPIOPError)
+      test.equal(err.apiErrorCode.code, '2000')
```
Comment (rmothilal) [inline test/unit/model/quotes.test.js]: Validate against the enum for errorCode from the Error library

File: test/unit/model/quotes.test.js [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+      }
+      Sinon.stub(db, 'newTransaction').returns(transaction)
+      Sinon.stub(db, 'getQuoteDuplicateCheck').returns({ hash: '85b6067dc6e271c53e2bbc2218e94187022677e80267f95ca28c80707b3009bc' })
+
+      await quotesModel.handleQuoteRequest(headers, quoteRequest)
+      test.fail('An error should be thrown')
+      test.end()
+    } catch (err) {
+      test.ok(err instanceof FSPIOPError)
+      test.equal(err.apiErrorCode.code, '3106')
```
Comment (rmothilal) [inline test/unit/model/quotes.test.js]: Validate against the enum for errorCode from the Error library

File: test/unit/model/quotes.test.js [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+      const quoteRequest = { }
+
+      Sinon.stub(db, 'getQuotePartyEndpoint').returns(null)
+
+      await quotesModel.forwardQuoteRequest(headers, 'test123', quoteRequest)
+      test.fail('Expected an error to be thrown')
+      test.end()
+    } catch (err) {
+      test.ok(err instanceof FSPIOPError)
+      test.equal(err.apiErrorCode.code, '3201')
```
Comment (rmothilal) [inline test/unit/model/quotes.test.js]: Validate against the enum for errorCode from the Error library

File: test/unit/model/quotes.test.js [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+      const quoteRequest = { }
+
+      Sinon.stub(db, 'getQuotePartyEndpoint').returns('http://invalid.com/dfsp2')
+
+      await quotesModel.forwardQuoteRequest(headers, 'test123', quoteRequest)
+      test.fail('Expected an error to be thrown')
+      test.end()
+    } catch (err) {
+      test.ok(err instanceof FSPIOPError)
+      test.equal(err.apiErrorCode.code, '1001')
```
Comment (rmothilal) [inline test/unit/model/quotes.test.js]: Validate against the enum for errorCode from the Error library

File: test/unit/model/quotes.test.js [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+      const quoteRequest = { }
+
+      Sinon.stub(db, 'getQuotePartyEndpoint').returns('http://invalidresponse.com/dfsp2')
+
+      await quotesModel.forwardQuoteRequest(headers, 'test123', quoteRequest)
+      test.fail('Expected an error to be thrown')
+      test.end()
+    } catch (err) {
+      test.ok(err instanceof FSPIOPError)
+      test.equal(err.apiErrorCode.code, '1001')
```
Comment (rmothilal) [inline test/unit/model/quotes.test.js]: Validate against the enum for errorCode from the Error library

File: src/model/quotes.js (hunk 1 above)
Comment (ndonnan) [reply src/model/quotes.js]: Now using NOT_IMPLEMENTED

File: src/model/quotes.js (hunk 2 above)
Comment (ndonnan) [reply src/model/quotes.js]: Done

File: src/model/quotes.js (hunk 3 above)
Comment (ndonnan) [reply src/model/quotes.js]: Done

File: src/model/quotes.js (hunk 4 above)
Comment (ndonnan) [reply src/model/quotes.js]: Done

**Pull request merged**